File: java/src/main/java/br/com/atlas/averbacao/security/KeyStore.java
package br.com.atlas.averbacao.security;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.MessageDigest;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.HexFormat;

public final class KeyStore {
    private KeyStore() {}

    public static BancoKeys generate() {
        try {
            KeyPairGenerator g = KeyPairGenerator.getInstance("Ed25519");
            KeyPair kp = g.generateKeyPair();
            String privatePem = writePem("PRIVATE KEY", kp.getPrivate().getEncoded());
            String publicPem = writePem("PUBLIC KEY", kp.getPublic().getEncoded());
            return new BancoKeys(privatePem, publicPem, keyidFromPublic(publicPem));
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static String keyidFromPublic(String publicPem) {
        try {
            byte[] hash = MessageDigest.getInstance("SHA-256").digest(publicPem.getBytes(StandardCharsets.UTF_8));
            return "banco_" + HexFormat.of().formatHex(hash).substring(0, 16);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static BancoKeys loadOrCreate(Path dataDir) {
        try {
            Files.createDirectories(dataDir);
            Path priv = dataDir.resolve("banco-private.pem");
            Path pub = dataDir.resolve("banco-public.pem");
            if (Files.exists(priv) && Files.exists(pub)) {
                String privatePem = Files.readString(priv);
                String publicPem = Files.readString(pub);
                return new BancoKeys(privatePem, publicPem, keyidFromPublic(publicPem));
            }
            BancoKeys keys = generate();
            Files.writeString(priv, keys.privateKey());
            Files.writeString(pub, keys.publicKey());
            return keys;
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static PrivateKey readPrivate(String pem) {
        try {
            return KeyFactory.getInstance("Ed25519").generatePrivate(new PKCS8EncodedKeySpec(decodePem(pem)));
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static PublicKey readPublic(String pem) {
        try {
            return KeyFactory.getInstance("Ed25519").generatePublic(new X509EncodedKeySpec(decodePem(pem)));
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static String newNonce() {
        byte[] n = new byte[16];
        new SecureRandom().nextBytes(n);
        return HexFormat.of().formatHex(n);
    }

    static String writePem(String type, byte[] der) {
        String b64 = Base64.getEncoder().encodeToString(der);
        StringBuilder sb = new StringBuilder();
        sb.append("-----BEGIN ").append(type).append("-----\n");
        for (int i = 0; i < b64.length(); i += 64) {
            sb.append(b64, i, Math.min(i + 64, b64.length())).append('\n');
        }
        sb.append("-----END ").append(type).append("-----\n");
        return sb.toString();
    }

    static byte[] decodePem(String pem) {
        String b64 = pem.replaceAll("-----BEGIN [^-]+-----", "")
            .replaceAll("-----END [^-]+-----", "")
            .replaceAll("\\s", "");
        return Base64.getDecoder().decode(b64);
    }
}

File: dotnet/Atlas.Averbacao/Security/KeyStore.cs
using System.Security.Cryptography;
using System.Text;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Generators;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.OpenSsl;
using Org.BouncyCastle.Pkcs;
using Org.BouncyCastle.Security;
using Org.BouncyCastle.X509;

namespace Atlas.Averbacao.Security;

public sealed record BancoKeys(string PrivateKey, string PublicKey, string Keyid);

public sealed record Identity(
    string AtlasPublicKey,
    string BancoPrivateKey,
    string BancoPublicKey,
    string BancoId,
    string Keyid);

public static class KeyStore
{
    public static BancoKeys Generate()
    {
        var gen = new Ed25519KeyPairGenerator();
        gen.Init(new Ed25519KeyGenerationParameters(new SecureRandom()));
        var pair = gen.GenerateKeyPair();
        var privatePem = WritePem("PRIVATE KEY", PrivateKeyInfoFactory.CreatePrivateKeyInfo(pair.Private).GetEncoded());
        var publicPem = WritePem("PUBLIC KEY", SubjectPublicKeyInfoFactory.CreateSubjectPublicKeyInfo(pair.Public).GetEncoded());
        return new BancoKeys(privatePem, publicPem, KeyidFromPublic(publicPem));
    }

    public static string KeyidFromPublic(string publicPem)
    {
        var hex = Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(publicPem))).ToLowerInvariant();
        return $"banco_{hex[..16]}";
    }

    public static BancoKeys LoadOrCreate(string dataDir)
    {
        Directory.CreateDirectory(dataDir);
        var privatePath = Path.Combine(dataDir, "banco-private.pem");
        var publicPath = Path.Combine(dataDir, "banco-public.pem");
        if (File.Exists(privatePath) && File.Exists(publicPath))
        {
            var privatePem = File.ReadAllText(privatePath);
            var publicPem = File.ReadAllText(publicPath);
            return new BancoKeys(privatePem, publicPem, KeyidFromPublic(publicPem));
        }
        var keys = Generate();
        File.WriteAllText(privatePath, keys.PrivateKey);
        File.WriteAllText(publicPath, keys.PublicKey);
        return keys;
    }

    public static AsymmetricKeyParameter ReadPrivate(string pem)
    {
        var obj = ReadPem(pem);
        return obj switch
        {
            AsymmetricCipherKeyPair pair => pair.Private,
            AsymmetricKeyParameter p => p,
            _ => throw new InvalidOperationException($"chave privada inesperada: {obj.GetType().Name}"),
        };
    }

    public static AsymmetricKeyParameter ReadPublic(string pem)
    {
        var obj = ReadPem(pem);
        return obj is AsymmetricCipherKeyPair pair ? pair.Public : (AsymmetricKeyParameter)obj;
    }

    private static object ReadPem(string pem)
    {
        using var reader = new StringReader(pem);
        var parsed = new PemReader(reader).ReadObject()
            ?? throw new InvalidOperationException("PEM inválido");
        return parsed;
    }

    private static string WritePem(string label, byte[] der)
    {
        var b64 = Convert.ToBase64String(der);
        var sb = new StringBuilder();
        sb.Append("-----BEGIN ").Append(label).AppendLine("-----");
        for (var i = 0; i < b64.Length; i += 64)
            sb.AppendLine(b64.Substring(i, Math.Min(64, b64.Length - i)));
        sb.Append("-----END ").Append(label).AppendLine("-----");
        return sb.ToString().Replace("\r\n", "\n");
    }
}

public sealed class IdentityState
{
    public Identity? Current { get; set; }
}
